### PR TITLE
feat(listeners): Add identifier to listeners

### PR DESCRIPTION
**Pull request by zmstone**
Listeners are internally identifiered by the listen-on tuple
which is not UI friendly when we have to find a listener by this
'signature'.

The listeners are actually named in configs, but the names are
discarded in the parsing functions.

In this PR, the configured names are kept and passed to emqx
internal for more user-friendly lookups.
The `emqx_ctl listeners` print out has been refactored to
display the identifier in format `<Protocol>[:<Transport>]:<Name>`.
e.g. `http:management` and `mqtt:tcp:internal`.
And the `emqx_ctl listeners stop` interface has been updated to
use the identifier as input.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked.:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## Backward Compatibility

For users, this PR only changes mqtt listeners' print-out lines.
Although the old stop CLI input is still supported,
the document as been updated to drop it: https://github.com/emqx/emqx-docs/pull/641

## More information

### CODE REVIEW

**Comment by zmstone** [comment]
A test case failure requires fix here:
https://github.com/emqx/emqx-ct-helpers/pull/31

**Comment by bgrosse-midokura** [comment]
@zmstone I can't seem to `emqx_ctl listeners stop <...>` any mqtt listeners with this. It works for `ws`, `wss`, `http:dashboard`, and `ssl`. But no combination of `mqtt:tcp`/`mqtt`/`tcp` and ip and/or port will stop plain mqtt listener: `RPC to emqx@127.0.0.1 error: {unknown_listener,...}`

**Comment by bgrosse-midokura** [comment]
@zmstone another observation:
```
~/p/emqx [refactor-assing-names-to-listeners ·] > _build/emqx/rel/emqx/bin/emqx_ctl listeners stop ssl 8883
Stop mqtt:ssl:external listener on 0.0.0.0:8883 successfully.
RPC to emqx@127.0.0.1 error: badarg
~/p/emqx [refactor-assing-names-to-listeners ·] > _build/emqx/rel/emqx/bin/emqx_ctl listeners |grep "^lis"
listener on mqtt:tcp:0.0.0.0:1883
listener on mqtt:tcp:127.0.0.1:11883
listener on http:dashboard:18083
listener on http:management:8081
listener on mqtt:ws:8083:8083
listener on mqtt:wss:8084:8084
```
Listener is actually stopped, but there is a strange additional `RPC to emqx@127.0.0.1 error: badarg` error message

**Comment by zmstone** [comment]
Hi @bgrosse-midokura 
It seems that your emqx node is built from an old commit (which has been force pushed).
This is from my tests:
```
> ./_build/emqx/rel/emqx/bin/emqx_ctl listeners
listener on mqtt:ssl:8883
  identifier      : mqtt:ssl:external
  acceptors       : 16
  max_conns       : 102400
  current_conn    : 0
  shutdown_count  : []
listener on mqtt:tcp:127.0.0.1:11883
  identifier      : mqtt:tcp:internal
  acceptors       : 4
  max_conns       : 1024000
  current_conn    : 0
  shutdown_count  : []
```
then 
```
> ./_build/emqx/rel/emqx/bin/emqx_ctl listeners stop mqtt:tcp:external
Stop mqtt:tcp:external listener on 0.0.0.0:1883 successfully.
RPC to emqx@127.0.0.1 error: badarg
```

I'm looking into the RPC error.

**Comment by zmstone** [comment]
@bgrosse-midokura Fixed the RPC error.

**Comment by bgrosse-midokura** [comment]
@zmstone it was this PR's commit, I just didn't understand the new "identifier" syntax, `mqtt:tcp:internal`. Thanks for explaining!

**Comment by bgrosse-midokura** [comment]
@zmstone I was wondering, if with this patch, couldn't the port part of the listener command be dropped entirely?

**Comment by zmstone** [comment]
> @zmstone I was wondering, if with this patch, couldn't the port part of the listener command be dropped entirely?

In the last commit, I added the new usage example for `stop` command.
The old `<Protocol> <Port>` is kept to be backward compatible.

For the restart API, yes, we can use just this identifier.

**Comment by bgrosse-midokura** [comment]
@zmstone got it, thanks!

File: lib-opensource/emqx_management/src/emqx_mgmt_cli.erl [hunk 1]
```diff
@@ -510,49 +510,64 @@ trace_off(Who, Name) ->
...
+stop_listener(false, Input) ->
+    emqx_ctl:print("No such listener ~p~n", [Input]);
+stop_listener(#{listen_on := ListenOn} = Listener, _Input) ->
+    ID = emqx_listeners:identifier(Listener),
+    ListenOnStr = emqx_listeners:format_listen_on(ListenOn),
+    case emqx_listeners:stop_listener(Listener) of
+        ok ->
+            emqx_ctl:print("Stop ~s listener on ~s successfully.~n", [ID, ListenOnStr]);
+        {error, Reason} ->
+            emqx_ctl:print("Failed to stop ~s listener on ~s - ~p~n.",
```
Comment (HJianBo) [inline lib-opensource/emqx_management/src/emqx_mgmt_cli.erl:568]:
```suggestion
            emqx_ctl:print("Failed to stop ~s listener on ~s - ~p.~n",
```

Furthermore, is it necessary to keep the formatting of the print the same as the 539 lines?

Comment (zmstone) [reply lib-opensource/emqx_management/src/emqx_mgmt_cli.erl:568]: there are test cases expecting the output.
and the format is also copied in listeners started by plugins.

**Review by HJianBo — APPROVED** [review]

**Pull request merged**